### PR TITLE
Issue 5257 - Ticket filters: Filters list and form

### LINKED ISSUE #5257: Titcket filters: base components
### Description
Product issue: https://github.com/3drepo/3D-Repo-Product-Team/issues/521
For advance ticket filtering we need to be able to show the list with all the properties from the templates select one and add the filter.

In this issue we should have to tackle two components: the **property selection list** and the base **criteria form**.

<details>
 <summary>Property selection list</summary>
 The property selection list includes all the properties for all the templates for which there are tickets in the container/federation.
For example if there are  two templates for a teamspace, one has "height" as number, the other has "colour" as string. And the only tickets that where created for that federation/container are the one with the "height" property then "colour" will not appear on the property selection list.
The property selection list also lets the user select other elements which are not necessarily "properties" per se, these are; 'title' (string), 'ticket id' (string), 'ticket template' (multi-select).
image

The properties within a module are depicted as modulename:propertyename  

</details>

###  Criteria form.
TBD

Filtering criteria depends on the type of property, the types are:
- **String** . Valid criterias:

   
- **Number**
- **Date**
- **Select** 

### Goals
TBD

- [ ] User journey 1
- [ ] User journey 2

### Tasks
- [ ] Task 2

### Related Resources

MockUp : `link to mockup`

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #5257 

#### Description
- The app now tells what properties can be used to filter tickets. These are listed in the menu that comes from the fennel. Here, only the non-used filters will be displayed (meaning, creating a new filter will remove it from that list)
- Clicking on a filter from that list shall open a ticket creation form. This is dumb, there's no validation and the number of available fields is only partly responsive (0, 1, or 3 fields)
- clicking on an existing filter should open the aforementioned form with pre-filled inputs

#### Test cases
Create new filters, edit them, remove them.
- The data displayed by the filters should be right,
- The available filters should always be up to date

### CODE REVIEW

File: frontend/src/v5/ui/components/viewer/cards/cardFilters/filtersSelection/tickets/ticketFiltersSelection.component.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+			onClose={onClose}
+			disabled={disabled}
+		>
+			<SearchContextComponent items={unusedFilters}>
+				<DrillDownList $visibleIndex={showFiltersList ? 0 : 1}>
+					<DrillDownItem $visible={showFiltersList}>
+						<SearchInput
+							placeholder={formatMessage({
+								id: 'viewer.card.tickets.filters.searchInputPlaceholder',
+								defaultMessage: 'Serach for property...',
```
Comment (The-Daniel) [inline frontend/src/v5/ui/components/viewer/cards/cardFilters/filtersSelection/tickets/ticketFiltersSelection.component.tsx:68]: Typo in "Search"

File: frontend/test/tickets/card/ticketsCard.store.spec.ts [hunk 2]
```diff
@@ -63,54 +64,92 @@ describe('Tickets: store', () => {
...
 		})
 
 		it('should reset the state', () => {
 			dispatch(TicketsCardActions.setSelectedTicket(ticketId));
 			dispatch(TicketsCardActions.setSelectedTemplate(templateId));
 			dispatch(TicketsCardActions.setSelectedTicketPin(pinId));
-			dispatch(TicketsCardActions.toggleCompleteFilter());
-			dispatch(TicketsCardActions.setTemplateFilters([templateId]));
-			dispatch(TicketsCardActions.setQueryFilters([query]));
+			dispatch(TicketsCardActions.resetFilters());
```
Comment (The-Daniel) [inline frontend/test/tickets/card/ticketsCard.store.spec.ts:141]: Shouldn't this upsert a filter?

File: frontend/test/test.helpers.ts [hunk 3]
```diff
@@ -59,7 +59,7 @@ export const createTestStore = () => {
 		return waitingActions.length === 0;
 	};
 
-	const mainReducer = combineReducers(reducers);
+	const mainReducer = createReducer({});
```
Comment (The-Daniel) [inline frontend/test/test.helpers.ts:62]: Why was this change necessary?

**Review by The-Daniel — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsList/ticketsListCard.component.tsx [hunk 4]
```diff
@@ -49,20 +50,21 @@ export const TicketsListCard = () => {
 				actions={(
 					<>
 						{!readOnly && (<NewTicketMenu />)}
+						<FilterSelection />
 						<EllipsisMenu>
 							<EllipsisMenuItemSwitch
 								icon={<PinIcon />}
-								title={formatMessage({ id: 'foobarbaz', defaultMessage: 'Show Pins' })}
+								title={formatMessage({ id: 'viewer.cards.tickets.showPins', defaultMessage: 'Show Pins' })}
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsList/ticketsListCard.component.tsx:57]: Yessssss

File: frontend/src/v5/ui/components/viewer/cards/cardFilters/cardFilters.types.ts [hunk 5]
```diff
@@ -16,4 +16,14 @@
  */
 
 export type CardFilterOperator = 'ex' | 'nex' | 'eq' | 'neq' | 'ss' | 'nss' | 'rng' | 'nrng' | 'gt' | 'gte' | 'lt' | 'lte';
-export type CardFilters = Record<string, any[]>;
\ No newline at end of file
+export type CardFilterType = 'text' | 'longText' | 'date' | 'pastDate' | 'sequencing' | 'oneOf' | 'manyOf' | 'boolean' | 'number' | 'ticketTitle' | 'ticketId' | 'template';
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/cards/cardFilters/cardFilters.types.ts:19]: That's for "created at". If you look at the template, its type is defined as such

File: frontend/src/v5/ui/routes/viewer/tickets/ticketsList/ticketItem/ticketItem.component.tsx [hunk 6]
```diff
@@ -85,9 +83,7 @@ export const TicketItem = ({ ticket }: TicketItemProps) => {
 			<FlexRow>
 				<FlexColumn>
 					<Title>
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsList/ticketItem/ticketItem.component.tsx:85]: Not sure about the future, but for now, we no longer have a query like before

[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/viewer/tickets/ticketsList/ticketItem/ticketItem.component.tsx:85]: Not sure about the future, but for now, we no longer have a query like before

File: frontend/test/test.helpers.ts (hunk 3 above)
Comment (Amantini1997) [reply frontend/test/test.helpers.ts:62]: All the v4 redux data was missing in the testing environment. This was to fix that

File: frontend/src/v5/store/tickets/tickets.types.ts [hunk 7]
```diff
@@ -212,8 +213,11 @@ export type OverridesDicts = {
...
-	complete: boolean,
-	templates: string[],
-	queries: string[],
+export type FilterTypesByProperty = {
+	[property in string]: {
+		[type in CardFilterType]?: BaseFilter;
+	};
+};
+export type FilterPropertiesByModule = {
+	[module in string]?: FilterTypesByProperty;
```
Comment (sanmont3drepo) [inline frontend/src/v5/store/tickets/tickets.types.ts:222]: I think the typing of this is a bit complex. Instead of this you could have an object that is more like
```
{
  'module.mycoolProp.number' : {operator: 'eq', values: [1]},
  'module.mycoolProp.string' : {operator: 'eq', values: ['brown', 'yellow']}
}
```
Maybe you could use templates literals to enforce a format?
https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html

File: frontend/src/v5/ui/components/viewer/cards/cardFilters/cardFilters.types.ts [hunk 8]
```diff
@@ -16,4 +16,14 @@
...
+export type CardFilterValue = string | number | Date;
+export type BaseFilter = { operator: CardFilterOperator, values: CardFilterValue[] };
+
+export type CardFilter = {
+	property: string,
+	type: CardFilterType,
+	filter: BaseFilter,
+};
+export type TicketCardFilter = CardFilter & { module: string };
+export type TicketCardFilterDescription = Omit<TicketCardFilter, 'filter'>;
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/components/viewer/cards/cardFilters/cardFilters.types.ts:29]: I dont think is necessary to have 3 types here. Just with one you can cater for all the cases:

```
export type CardFilter = {
	property: string,
	type: CardFilterType,
	filter?: BaseFilter,
	module?: string,
};
```

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/tickets/card/ticketsCard.redux.ts [hunk 9]
```diff
@@ -96,16 +93,19 @@ export const setPinToDrop = (state: ITicketsCardState, { pinToDrop }: SetPinToDr
...
 
-export const toggleCompleteFilter = (state: ITicketsCardState) => {
-	state.filters.complete = !state.filters.complete;
+export const deleteFilter = (state: ITicketsCardState, { filter }: DeleteFilterAction) => {
+	const path = getFilterPath(filter);
+	// delete filter and recursiverly delete parent entry if empty
+	do {
+		unset(state.filters, path);
+		path.pop();
+	} while (path.length && isEmpty(get(state.filters, path)));
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/tickets/card/ticketsCard.redux.ts:108]: The new logic will no longer use this code

File: frontend/src/v5/ui/components/viewer/cards/cardFilters/filtersSelection/tickets/ticketFiltersSelection.styles.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+		.MuiInputAdornment-positionEnd {
+			margin: 0;
+		}
+	}
+`;
+
+export const DrillDownItem = styled.div<{ $visible: boolean }>`
+	transition: opacity height .2s;
+	${({ $visible }) => $visible ? css`
+		height: 100%;
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/viewer/cards/cardFilters/filtersSelection/tickets/ticketFiltersSelection.styles.ts:45]: I changed the transition so that it applies to the max-height

**Review by sanmont3drepo — APPROVED** [review]

**Review by The-Daniel — APPROVED** [review]
